transport/rif_comms/node.py

```diff
@@ -1,51 +1,339 @@
+            # Ignore when stopped
+            return False
+
+        # TODO validate signature of the message
+
+        messages = list()  # TODO validate_and_parse_message(event["content"]["body"], peer_address)
+
+        if not messages:
+            return False
+
+        self.log.info(
+            "Incoming messages",
+            messages=messages,
+            topic_id=pex(topic_id),
+        )
+
+        for message in messages:
+            if not isinstance(message, (SignedRetrieableMessage, SignedMessage)):
+                self.log.warning("Received invalid message", message=message)
+            if isinstance(message, Delivered):
+                self._receive_delivered(message)
+            elif isinstance(message, Processed):
+                self._receive_message(message)
+            else:
+                assert isinstance(message, SignedRetrieableMessage)
+                self._receive_message(message)
+
+        return True
+
+    def _receive_delivered(self, delivered: Delivered):
+        self.log.info(
+            "Delivered message received", sender=pex(delivered.sender), message=delivered
+        )
+
+        assert self._raiden_service is not None
+        self._raiden_service.on_message(delivered)
+
+    def _receive_message(self, message: Union[SignedRetrieableMessage, Processed]):
+        assert self._raiden_service is not None
+        self.log.info(
+            "RIF Comms Message received",
+            node=pex(self._raiden_service.address),
+            message=message,
+            sender=pex(message.sender),
+        )
+
+        try:
+            delivered_message = Delivered(delivered_message_identifier=message.message_identifier)
+            self._raiden_service.sign(delivered_message)
+
+            queue_identifier = QueueIdentifier(
+                recipient=message.sender, channel_identifier=CHANNEL_IDENTIFIER_GLOBAL_QUEUE
+            )
+            self.send_message(*TransportMessage.wrap(queue_identifier, delivered_message))
+            self._raiden_service.on_message(message)
+
+        except (InvalidAddress, UnknownAddress, UnknownTokenAddress):
+            self.log.warning("Exception while processing message", exc_info=True)
+            return
 
     def start_health_check(self, address: Address):
-        raise NotImplementedError
+        self.log.debug("Healthcheck", peer_address=pex(address))
 
     def whitelist(self, address: Address):
-        raise NotImplementedError
+        self.log.debug("Whitelist", peer_address=pex(address))
 
     def link_exception(self, callback: Any):
-        raise NotImplementedError
+        self.greenlet.link_exception(callback)
 
     def join(self, timeout=None):
-        raise NotImplementedError
+        self.greenlet.join(timeout)
+
+    def listen_messages(
+        self
+    ):
+        """
+        Iterate over the Notification stream and blocks thread to receive messages
+        """
+        for notification_stream in self._our_topic:
```

Comment:
i am not entirely sure how this `for` will work in a greenlet. do you have an inkling of this?

Reply:
The for in loop is the standar way to listen to `streams `on GRPC 

- https://stackoverflow.com/questions/47831895/how-do-i-handle-streaming-messages-with-python-grpc
- https://github.com/grpc/grpc/blob/v1.20.0/examples/python/route_guide/route_guide_client.py#L96

Regarding the greenlet, i think the for loop blocks the current thread until a message arrives, therefore we need a greenlet running this background task in order to not block the ppal thread